### PR TITLE
[RN][Codegen] Add docs for advanced iOS usage

### PR DESCRIPTION
**Pull request by cipolleschi**
Update docs for Codegen for 0.76 and 0.77

### CODE REVIEW

**Review by cortinico — APPROVED** [review]

File: docs/the-new-architecture/using-codegen.md [hunk 1]
```diff
@@ -46,7 +62,13 @@ You can add this snippet to your app and customize the various fields:
   - `components:` use this value if you only need to generate code for Native Fabric Components.
   - `all`: use this value if you have a mixture of components and modules.
 - `jsSrcsDir`: this is the root folder where all your specs live.
-- `android.javaPackageName`: this is an android specific setting to let **Codegen** generate the files in a custom package.
+- `android.javaPackageName`: this is an Android specific setting to let **Codegen** generate the files in a custom package.
+- `ios`: the `ios` field is an object that can be used by app developers and library maintainers to provide some advanced functionalities. All the following fields are **optional**.
+  - `ios.modulesConformingToProtocol`: React Native offers some protocols that native modules can implement to customize some behaviors. This fields allow to define the list of modules that conforms to those protocols. These modules will be injected in the React Native runtime when the app starts.
```
Comment (coolsoftwaretyler) [inline docs/the-new-architecture/using-codegen.md:67]:
```suggestion
  - `ios.modulesConformingToProtocol`: React Native offers some protocols that native modules can implement to customize some behaviors. These fields allow you to define the list of modules that conform to those protocols. These modules will be injected in the React Native runtime when the app starts.
```

**nitpick:** couple typos

File: docs/the-new-architecture/using-codegen.md [hunk 2]
```diff
@@ -46,7 +62,13 @@ You can add this snippet to your app and customize the various fields:
...
+    - `ios.modulesConformingToProtocol.RCTImageURLLoader`: list of iOS native modules that implements the [`RCTImageURLLoader` protocol](https://github.com/facebook/react-native/blob/00d5caee9921b6c10be8f7d5b3903c6afe8dbefa/packages/react-native/Libraries/Image/RCTImageURLLoader.h#L26-L81).
```
Comment (coolsoftwaretyler) [inline docs/the-new-architecture/using-codegen.md:68]:
```suggestion
    - `ios.modulesConformingToProtocol.RCTImageURLLoader`: list of iOS native modules that implement the [`RCTImageURLLoader` protocol](https://github.com/facebook/react-native/blob/00d5caee9921b6c10be8f7d5b3903c6afe8dbefa/packages/react-native/Libraries/Image/RCTImageURLLoader.h#L26-L81).
```

**nitpick:** typo

File: docs/the-new-architecture/using-codegen.md [hunk 3]
```diff
@@ -46,7 +62,13 @@ You can add this snippet to your app and customize the various fields:
...
+    - `ios.modulesConformingToProtocol.RCTImageDataDecoder`: list of iOS native modules that implements the [`RCTImageDataDecoder` protocol](https://github.com/facebook/react-native/blob/00d5caee9921b6c10be8f7d5b3903c6afe8dbefa/packages/react-native/Libraries/Image/RCTImageDataDecoder.h#L15-L53).
```
Comment (coolsoftwaretyler) [inline docs/the-new-architecture/using-codegen.md:69]:
```suggestion
    - `ios.modulesConformingToProtocol.RCTImageDataDecoder`: list of iOS native modules that implement the [`RCTImageDataDecoder` protocol](https://github.com/facebook/react-native/blob/00d5caee9921b6c10be8f7d5b3903c6afe8dbefa/packages/react-native/Libraries/Image/RCTImageDataDecoder.h#L15-L53).
```

**nitpick:** typo

File: docs/the-new-architecture/using-codegen.md [hunk 4]
```diff
@@ -46,7 +62,13 @@ You can add this snippet to your app and customize the various fields:
...
+    - `ios.modulesConformingToProtocol.RCTURLRequestHandler`: list of iOS native modules that implements the [`RCTURLRequestHandler` protocol](https://github.com/facebook/react-native/blob/00d5caee9921b6c10be8f7d5b3903c6afe8dbefa/packages/react-native/React/Base/RCTURLRequestHandler.h#L11-L52).
```
Comment (coolsoftwaretyler) [inline docs/the-new-architecture/using-codegen.md:70]:
```suggestion
    - `ios.modulesConformingToProtocol.RCTURLRequestHandler`: list of iOS native modules that implement the [`RCTURLRequestHandler` protocol](https://github.com/facebook/react-native/blob/00d5caee9921b6c10be8f7d5b3903c6afe8dbefa/packages/react-native/React/Base/RCTURLRequestHandler.h#L11-L52).
```

**nitpick:** typo

File: docs/the-new-architecture/using-codegen.md [hunk 5]
```diff
@@ -46,7 +62,13 @@ You can add this snippet to your app and customize the various fields:
...
+  - `ios.componentProvider`: this field is a map used to generate the association between a custom JS React component and the native class that implements it. The key of the map is the JS name of the component (for example `TextInput`), and the value is the iOS class that implements the component (for example `RCTTextInput`).
```
Comment (coolsoftwaretyler) [inline docs/the-new-architecture/using-codegen.md:71]: This is so cool - I'm excited to see this in the docs.

File: website/versioned_docs/version-0.76/the-new-architecture/using-codegen.md [hunk 6]
```diff
@@ -46,7 +59,12 @@ You can add this snippet to your app and customize the various fields:
   - `components:` use this value if you only need to generate code for Native Fabric Components.
   - `all`: use this value if you have a mixture of components and modules.
 - `jsSrcsDir`: this is the root folder where all your specs live.
-- `android.javaPackageName`: this is an android specific setting to let **Codegen** generate the files in a custom package.
+- `android.javaPackageName`: this is an Android specific setting to let **Codegen** generate the files in a custom package.
+- `ios`: the `ios` field is an object that can be used by app developers and library maintainers to provide some advanced functionalities. All the following fields are **optional**.
+  - `ios.modulesConformingToProtocol`: React Native offers some protocols that native modules can implement to customize some behaviors. This fields allow to define the list of modules that conforms to those protocols. These modules will be injected in the React Native runtime when the app starts.
```
Comment (coolsoftwaretyler) [inline website/versioned_docs/version-0.76/the-new-architecture/using-codegen.md:64]:
```suggestion
  - `ios.modulesConformingToProtocol`: React Native offers some protocols that native modules can implement to customize some behaviors. These fields allow you to define the list of modules that conforms to those protocols. These modules will be injected in the React Native runtime when the app starts.
```

**nitpick:** typos

**Review by coolsoftwaretyler — COMMENTED** [review]
Hey @cipolleschi - thanks for adding these docs. Such a cool feature of codegen.

I left a couple of typo fix suggestions for you, and I have one general suggestion. I learned about this field through this PR into react-native-cameraroll: https://github.com/react-native-cameraroll/react-native-cameraroll/pull/632/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R122-R126

If you wanted to add examples of why or how to use these fields, I feel like that PR made a lot of sense to me (enable handling certain file handles that `RCTImageURLLoader` had dropped support for due to the lean core initiative.

Either way, I appreciate your time on this one. :)

File: website/versioned_docs/version-0.76/the-new-architecture/using-codegen.md [hunk 7]
```diff
@@ -34,6 +34,19 @@ npx @react-native-community/cli@latest init SampleApp --version 0.76.0
     "jsSrcsDir": "<source_dir>",
     "android": {
       "javaPackageName": "<java.package.name>"
+    },
+    "ios": {
+      "modulesConformingToProtocol": {
+        "RCTImageURLLoader": [
+          "<iOS-module-conforming-to-RCTImageURLLoader>",
```
Comment (elicwhite) [inline website/versioned_docs/version-0.76/the-new-architecture/using-codegen.md:41]: It's not clear to me what should go here. Is it the module name? `MyImageLoader.mm`? The absolute path to the file? Relative path from project root? Just the class name `MyImageLoader`? 

Is that clear from somewhere else in the markdown file? If not, consider adding an example to the markdown and clarify in the text.

Comment (cipolleschi) [reply website/versioned_docs/version-0.76/the-new-architecture/using-codegen.md]: Updated both the JSON syntax and the description of those fields below, referencing the `react-native-camera-roll` as a real life example.

File: docs/the-new-architecture/using-codegen.md [hunk 8]
```diff
@@ -34,6 +34,27 @@ npx @react-native-community/cli@latest init SampleApp --version 0.76.0
...
+          // "RNCPHAssetLoader",
+        ],
+        "RCTURLRequestHandler": [
+          "<iOS-class-conforming-to-RCTURLRequestHandler>",
+          // example from react-native-camera-roll: https://github.com/react-native-cameraroll/react-native-cameraroll/blob/8a6d1b4279c76e5682a4b443e7a4e111e774ec0a/package.json#L118-L127
+          // "RNCPHAssetUploader",
+        ],
+        "RCTImageDataDecoder": [
+          "<iOS-class-conforming-to-RCTImageDataDecoder>",
+          // we don't have a good example for this, but it works in the same way. Pass the name of the class that implements the RCTImageDataDecoder. It must be a Native Modules.
```
Comment (okwasniewski) [inline docs/the-new-architecture/using-codegen.md:52]:
```suggestion
          // we don't have a good example for this, but it works in the same way. Pass the name of the class that implements the RCTImageDataDecoder. It must be a Native Module.
```

**Pull request merged**